### ISSUE TITLE
Extend metrics endpoint

### ISSUE THREAD

**Issue by hifabienne** [zitadel/zitadel#4458]
We already have a metrics endpoint but this is only about how many request we had on each endpoint. We should extend the endpoint with the following metrics:

**Acceptance criteria**
- [ ] Login count
- [ ] Failed logins
- [ ] Logins per application
- [ ] Failed logins per application
- [ ] Org count
- [ ] Project count
- [ ] Application count
- [ ] User count
- [ ] User count per state
- [ ] Creation Timestamp of the first project
- [ ] Creation Timestamp of the first application

**Comment by hifabienne** [comment]
@stebenz @livio-a I have added some more insights about why I added which metrics/reports and what the usecase is

**Comment by livio-a** [comment]
> @stebenz @livio-a I have added some more insights about why I added which metrics/reports and what the usecase is

perfect. this gives an good overview and can create multiple tasks out of it

**Comment by hifabienne** [comment]
@eliobischof @livio-a @stebenz can you have a look at it how we should split and make subtasks?

**Comment by eliobischof** [comment]
Backend:
To me, all these use cases look very much like reporting, where CQRS shines. Especially the selectable timeframe parts.
-> Projections.

Frontend:
Maybe some standard dashboards with nice histograms?

Splitting:
Maybe for this one, we could consider splitting the parts along the process, so it's easier to ensure consistent implementation: "data model design and data flows", "UI and API design", "backend implementation", "frontend implementation".

**Comment by fforootd** [comment]
Just to get the discussion started here. In my mind all our events can be directly converted into metrics.
On the other hand I see "reports" which to me is just a query. So let me dump my thoughts here 😁 

## Events

The full payload that happend

## Metrics

The arithmetic something happend

- Arithmetic's of the events we stored (count, sum, ..) over a given time
- Arithmetic's of the access logs of our system
- Arithmetic's of the execution logs from actions
- Arithmetic's of underlying components like CPU/Memory usage, Go GC, ...

## Query

The questions I have

- I want to have the amount of failed login per client/project
- I want to have see the top 10 users with auth failures
- I want to see the most active IdP
- I want to see the last execution logs
- I want to see the last 5 events of a user

## Dashboard

The visualization of the queries

## Discussion

I think something like https://grafana.com/docs/loki/latest/logql/ would be interesting since it supports metric query https://grafana.com/docs/loki/latest/logql/metric_queries/#metric-queries and has logical operations

My 2 cents are, lets query events/access/execution with logql and create some low level metrics like cpu/memory/... as promql

More benefits of this route would be that we could allow our users to easy onboard with tools that understand logql/promql

What do you think?

**Comment by eliobischof** [comment]
So the idea is to provide

- A scrapable endpoint for binary runtime metrics (time series are stored in an external system like Prometheus)
- A promql compatible API (Time series are stored in ZITADEL projections and/or the ZITADEL eventstore is queried directly

Did I get that right?

**Comment by fforootd** [comment]
> So the idea is to provide
> 
> * A scrapable endpoint for binary runtime metrics (time series are stored in an external system like Prometheus)
> * A promql compatible API (Time series are stored in ZITADEL projections and/or the ZITADEL eventstore is queried directly)
> * A logql compatible API (ZITADEL logstores and eventstore are queried directly)
> 
> Did I get that right?

Yes thats what i think could work. I think in theory all metrics could also be published as metrics endpoint, but i am not sure of the value at that end.

**Comment by eliobischof** [comment]
> > So the idea is to provide
> > 
> > * A scrapable endpoint for binary runtime metrics (time series are stored in an external system like Prometheus)
> > * A promql compatible API (Time series are stored in ZITADEL projections and/or the ZITADEL eventstore is queried directly)
> > * A logql compatible API (ZITADEL logstores and eventstore are queried directly)
> > 
> > Did I get that right?
> 
> Yes thats what i think could work. I think in theory all metrics could also be published as metrics endpoint, but i am not sure of the value at that end.

AFAIK, a prom scraper compatible metrics endpoint exposes blank values which are then mapped to a timestamp in an external time series database. But yes, I think we can provide each metric like that too, I guess.

**Comment by fforootd** [comment]
> > > So the idea is to provide
> > > 
> > > * A scrapable endpoint for binary runtime metrics (time series are stored in an external system like Prometheus)
> > > * A promql compatible API (Time series are stored in ZITADEL projections and/or the ZITADEL eventstore is queried directly)
> > > * A logql compatible API (ZITADEL logstores and eventstore are queried directly)
> > > 
> > > Did I get that right?
> > 
> > 
> > Yes thats what i think could work. I think in theory all metrics could also be published as metrics endpoint, but i am not sure of the value at that end.
> 
> AFAIK, a prom scraper compatible metrics endpoint exposes blank values which are then mapped to a timestamp in an external time series database. But yes, I think we can provide each metric like that too, I guess.

Yes it is like that, hence my promql to make queries happen

**Comment by mffap** [comment]
@hifabienne we should also have a metric "MAU" defined as users who have logged in or refreshed their token in a given month (or period)

Please also consider that it would be useful to count both Service Accounts and Human Users. Split and Aggregate metrics by user type would be a plus.

This would help translate MAU or Users to Authenticated Requests per Instance.

**Comment by kraney** [comment]
For successful logins, it would be useful to also count the login method - local password, IdP, was MFA or passwordless used, JWT vs Bearer token, etc. And for the token endpoint, useful to know counts of when the refresh token was used, etc.

Also I would say that the metrics are most useful exported as OpenTelemetry, so that they can be rolled into other monitoring solutions. Not just available from within Zitadel.
